doc/Frontend/syntax.org

```diff
@@ -0,0 +1,126 @@
+
+  function ::= <non-capital-name> <name>+ = <expression>
+
+  ; if we only have usage in arrows, we should allow a way to have
+  ; usage information of variables as well!
+  signature ::= <non-capital-name> <type-signature>
+              | <non-capital-name> <usage> <type-signature>
+
+  ;;; Types ==========================================================
+
+  type ::= <data-declaration>
+         | <type-declaration>
+         | <alias-declaration>
+
+  ; Should we let there be a refinement in adt type declaration?
+  data-declaration  ::= type <non-capital-name> : <arrow-type> = <adt>
+                      | type <non-capital-name>                = <adt>
+
+  type-declaration  ::= type <non-capital-name> = <name> <type-refine>
+  alias-declaration ::= type <non-capital-name> = <type-refine>
+
+  adt ::= <sum>*
+        | <product>
+
+
+  ; | is special in bnf
+  sum ::= \| <name>
+        | \| <name> <product>
+
+  product ::= { <name-type-semi>+ <name-type>* }
+            | { <name-type-semi>* }
+            | <arrow-type>
+
+  name-type      ::= <name> <type-signature>
+  name-type-semi ::= <name-type> <semi>
+
+  ; should we move ()'s somewhere else?
+  arrow-type ::= <type-refine>
+               | <name> : <type-refine> <arrow> <arrow-type>
+               | <type-refine>          <arrow> <arrow-type>
+               | ( <arrow-type> )
+
+  type-signature ::= : <arrow-type>
+
+  ; { } here are a refinement type!
+  type-refine ::= <type-name> { <expression> }
+                | <type-name>
+
+  ;;; Arrows =========================================================
+
+
+  arrow ::= -\> ; ω usage by default? This should be the most common arrow
+          | -o  ; linear usage
+          | -|  ; ??? some kind of usage
+          | -<usage>-\>
+          | --<usage>--\> ; maybe remove this one?
+
+  ;;; Modules ========================================================
+
+  ; This doubles as our import
+  module-open ::= open <module-name>
+
+  ; We are going to make modules a bit more interesting, syntax wise
+  ; imagine modules were functions with capital name to deliminate
+  ; thus module signatures have the same signature look as functions
+  ; however the end of this signature is quite a bit different than normal
+  ; functions, Maybe has { }, maybe has end? I'm not really sure yet
+
+  ; Don't need end if we are indent sensitive!
+  module ::= <module-name> <module-name>+ = <top-level>+ end
+
+  ;;; Types Classes ==================================================
+
+  ; Need end if we are indent sensitive!
+  type-class ::= class
+
+  ; Need end if we are indent sensitive!
+  type-class-instance ::= instance <..> where
+
+  ;;; Expressions ====================================================
+  expression ::=
+
+  usage ::= <expression>
+
+
+  record-access ::= <name>.<name>
+
+  module-lookup ::= <module-name>.<name>
+
+  ; we should remove either if or cond!?
+  if   ::= if   <cond-logic>*
```

Comment:
I'm not sure if we even want to special-case this for now. Ideally we can have some "syntax" pragma like Idris.

Reply:
Well we need some kind of if semantics.


notably

```ocaml
if b
then a
else c
```

is more special cased than

```haskell
if | b = a
   | else = c
```

Because the latter form already exists in the form of

```haskell
f b =
  | foo b = a
  | otherwise = c
```

Reply:
Right right, guard syntax seems reasonable to me.

Reply:
One note is that we'll want nice exhaustiveness checking (this might need to use proofs in the context, it's not possible in general of course).